### ISSUE TITLE
Boost a single field

### ISSUE THREAD

**Issue by meltuhamy** [olivernn/lunr.js#263]
Is it possible to boost a field in the search (v2)? 

Example:
On the demo page, https://olivernn.github.io/moonwalkers/ , if I search two fields like this: `name:pete^10 body:pete`, I expect the result which has a match on the **name** field to come first. However, the result with the match in the body comes first.

**Comment by j1m1lo** [comment]
I also tried to achieve field boosting after the upgrade, and ran into the same issue.

However, I then checked the observed issues also with my old setup running lunr 0.7.2 - and essentially found the same unexpected behaviour: Hits in the body kind of dominate the results, and boosting even using large values seems to have only a marginal effect on scores. It just was/is more obscure there, because only lunr 2.0 gives you the score details in the result metadata.

Could this maybe be caused by the fact that body content is (a lot) larger than other fields' content? @olivernn it would be great if you could shed some light onto this. Thanks!

**Comment by chriskarkowsky** [comment]
I am facing the same issue at the moment. Heavily weighing one field over others should take precedence in the index.

In the example, if you ran "name:pete^10 body:flight" or simply "name:pete^10 flight", I would expect the Pete Conrad record to be first.

**Comment by olivernn** [comment]
Yeah, the boost is definitely not working as intended 😞

I'll take a look and see what is going wrong, thanks for letting me know.

**Comment by olivernn** [comment]
I've got a good idea whats going wrong here and I think I've got a solution. Specifically it is to do with how lunr is treating documents, currently there is just a single vector space model for the entire document. With field search it really needs to have a vector space of each individual field. I have a rough implementation of this locally and it performs much better with field based searches.

I'm just trying to figure out a way of _testing_ the relevancy of search results in an automated, repeatable manner. This is tricky to get right. I've found some [resources](https://cwiki.apache.org/confluence/display/ORP/ExistingCollections) from a now defunct [apache project](https://cwiki.apache.org/confluence/display/ORP/Index) that might be useful in assessing how good the relevancy of lunr is before and after this change. Basically, I want to be able to see the impact on changing the scoring algorithms within lunr.

Anyway, long story short, I think I've got a good solution to this issue, I just need to make sure it doesn't negatively impact any other kind of searches.

**Comment by Andargor** [comment]
I've also noticed some examples with people boosting the index so that a field (e.g. "title") will have more weight by default, example: this.field('title', {boost: 10});

However, I looked at builder.js and there is no support for that, the field is simply pushed, and there are no options such as boost. Would be nice to include this.

**Comment by olivernn** [comment]
@Andargor this part of the interface has changed since 2.0.0, this is covered in the [upgrading guide](https://lunrjs.com/guides/upgrading.html#index-definition).

That said, with the changes discussed earlier in this thread, it _should_ be possible to bring back this feature, allowing build time boosts for certain fields again.

**Comment by Andargor** [comment]
@olivernn Ok, I see that now, thanks. I've tried the new method for example to better score the search term if found in the title field than the body field, and as outlined here I am not seeing any measurable difference in scoring. I also tried something along the lines of "title:term^10 body:term" to no avail.

My current workaround is placing the results with title metadata first in the list when assembling HTML, but ideally building in the boost in the index would be preferable.

**Comment by olivernn** [comment]
I've just published an alpha build of [lunr 2.1](https://github.com/olivernn/lunr.js/tree/2.1) with the changes to improve the field based search and boosting. If you have time please try it out to see if it solves the issues you were seeing.

I have tested it locally with the moonwalkers site and the queries mentioned in this issue now return results as expected.

You can install the alpha with the following:

```javascript
{
  "dependencies": {
    "lunr": "alpha"
  }
}
```

If all looks right I'm hoping to release a non-alpha build in the next couple of days.

**Comment by j1m1lo** [comment]
I've tried around with the alpa a bit.

Search results are much better now for my corpus and test queries. That is, without field boosting.

I'm still fiddling around with field boosting. In many cases, there is no effect on result scores, or at least not in the way I would expect. Without a public demo, this is needless to discuss though. I'll continue experimenting, anyway.

Finally, unboosted search works so well I am not sure whether I even need field boosting, anymore.

**Comment by olivernn** [comment]
@j1m1lo yeah, I think the issue is that they are really term boosts _within_ a field not across fields. Getting them to work across fields now might be tricky. I might need to rethink these changes. What I have in mind _shouldn't_ affect the un-boosted search results that seem to be working much better now. Thanks for helping with the testing of this!

**Comment by olivernn** [comment]
I've released [2.0.4](https://github.com/olivernn/lunr.js/releases/tag/v2.0.4) which includes a fix that was probably affecting some non-boosted search results. It actually fixes the query at the beginning of the issue too.

That said I need to do some more investigation/testing into how field boosts currently work and how they should.

**Comment by Andargor** [comment]
@olivernn Good work! It was a bit unfair of me to sneak in a feature request in someone else's bug report, even if it was loosely related, so feel free to close the issue. I'm in the middle of a deadline right now and have not tested your alpha. Do you want me to open a separate feature request?

**Comment by olivernn** [comment]
@Andargor yeah, please open another issue regarding the build time field boosts, just makes it easier for me to keep track of things.

**Comment by nkuehn** [comment]
Feedback to the 2.1 alpha:  I just experimentally applied it to a software documentation site with three indexed fields (title, description, body).   

The results are pretty promising, even without passing any boost information at all. The index behaves more "intuitive" now, clearly preferring matches in title or description over body matches.

**Comment by olivernn** [comment]
@nkuehn thanks for taking the time to test out the alpha. I've also released 2.0.4 which _should_ include most of the improvements to result relevancy without having any major structural changes to the index, and hence no minor version bump.

**Comment by nkuehn** [comment]
you're welcome.  I tried 2.0.4, too but it does not yield good enough results without boosting at query time.   Only 2.1 had ootb-in behavior that did not require boosting.

**Comment by meltuhamy** [comment]
2.1 boosting worked for me too. No luck with 2.0.4

**Comment by olivernn** [comment]
I've just pushed a new alpha version (2.1.0-alpha.2) which has some further improvements for field based boosting, results should be better where a term is boosted in one field, but not another.

If everything seems fine I'm hoping to publish a non-alpha in the next couple of days.

**Comment by chriskarkowsky** [comment]
I am having some good luck with the newer alpha build.  I might need to change my data structure a bit based on my issue here with array data: https://github.com/olivernn/lunr.js/issues/266, but it shouldn't be too big of an issue. 

Thank you!

**Comment by olivernn** [comment]
I've just pushed 2.1.0 with the changes discussed to improve boosting and field based search.
